source/common/formatter/substitution_formatter.h

```diff
@@ -243,6 +244,251 @@ template <class FormatterContext> class FormatterBaseImpl : public FormatterBase
   std::vector<FormatterProviderBasePtr<FormatterContext>> providers_;
 };
 
+// Helper class to write value to output buffer in JSON style.
+// NOTE: This helper class has duplicated logic with the Json::Streamer class but
+// provides lower level of APIs to operate on the output buffer (like control the
+// delimiters). This is designed for special scenario of substitution formatter and
+// is not intended to be used by other parts of the code.
+class JsonStringSerializer {
+public:
+  using OutputBufferType = Json::StringOutput;
+  explicit JsonStringSerializer(std::string& output_buffer) : output_buffer_(output_buffer) {}
+
+  // Methods that be used to add JSON delimiter to output buffer.
+  void addMapBeginDelimiter() { output_buffer_.add(Json::Constants::MapBegin); }
+  void addMapEndDelimiter() { output_buffer_.add(Json::Constants::MapEnd); }
+  void addArrayBeginDelimiter() { output_buffer_.add(Json::Constants::ArrayBegin); }
+  void addArrayEndDelimiter() { output_buffer_.add(Json::Constants::ArrayEnd); }
+  void addElementsDelimiter() { output_buffer_.add(Json::Constants::Comma); }
+  void addKeyValueDelimiter() { output_buffer_.add(Json::Constants::Colon); }
+
+  // Methods that be used to add JSON key or value to output buffer.
+  void addString(absl::string_view value) { addSanitized(R"(")", value, R"(")"); }
+  void addNumber(double d) {
+    if (std::isnan(d)) {
+      output_buffer_.add(Json::Constants::Null);
+    } else {
+      Buffer::Util::serializeDouble(d, output_buffer_);
+    }
+  }
+  void addNumber(uint64_t i) { output_buffer_.add(absl::StrCat(i)); }
+  void addNumber(int64_t i) { output_buffer_.add(absl::StrCat(i)); }
```

Comment:
https://stackoverflow.com/questions/13502398/json-integers-limit-on-size is of interest; TL;DR the json spec does not describe this limit. It might be that precision is lost if this is read by JavaScript, but a C++ json reader is free to retain the precision. There is an interoperability note in RFC7159 referencing this, but my feeling is we should not clamp to that value, and leave it to the JSON reader to drop precision if needed.

Reply:
...although putting a comment in here referencing the json spec and RFC about numbers would make sense -- we should also add that to json_streamer.h where all this is sadly duplicated :)

Reply:
Not sure if it necessary to add additional comments for this.

Although all json numbers are float in definition. But in the actual implementation, most json lib provide the integer support.

Maybe some lib may lost precision when loading the output, but we basically have no better choice anyway?

Reply:
I agree with your reasoning, and think it would be good to write a comment so that readers are on the same page.

Reply:
>  we should also add that to json_streamer.h where all this is sadly duplicated :)

I think we could let the Streamer inherit from here's serializer in the future, like my very initial suggestion. They basically are different levels of abstraction.

Reply:
comments are done.

Reply:
the serializer should definitely not inherit from the formatter. But a common base-class (or even just helper methods) could be factored out.

Reply:
> the serializer should definitely not inherit from the formatter. But a common base-class (or even just helper methods) could be factored out.

we can move the new serializer here out from formatter. I think this doesn't matter. we can give a try at other PR to see if it make sense.